crates/fuel-streams-core/src/stream.rs

```diff
@@ -0,0 +1,230 @@
+mod error;
+
+use std::fmt::Debug;
+
+use async_nats::{
+    jetstream::{
+        consumer::AckPolicy,
+        kv,
+        stream::{self, LastRawMessageErrorKind},
+    },
+    Message,
+};
+use async_trait::async_trait;
+pub use error::StreamError;
+use fuel_data_parser::{
+    DataParser,
+    DataParserDeserializable,
+    DataParserSerializable,
+    NatsFormattedMessage,
+};
+use fuel_streams_macros::subject::IntoSubject;
+use futures_util::StreamExt;
+use tokio::sync::OnceCell;
+
+use crate::{nats::types::*, prelude::NatsClient};
+
+#[async_trait]
+pub trait StreamEncoder:
+    Debug
+    + Clone
+    + Send
+    + Sync
+    + DataParserSerializable
+    + DataParserDeserializable
+    + 'static
+{
+    async fn encode(&self, subject: &str) -> Vec<u8> {
+        Self::data_parser()
+            .to_nats_payload(subject, self)
+            .await
+            .expect("Streamable must encode correctly")
```

Comment:
I feel we should handle the error here rather than unwrap it ?

Reply:
Since we use explicit error handling, I am trying to reduce the complexity of our systems by making anomalies obvious with unwraps and useful errors should remain handled explicitly. What do you think?

Reply:
to nats_payload throws errors that come about compaction/serialization etc. I believe you could at least cathc it and print it here with error! so we can see it before panicking. I believe the error should be handled here, but do not insist on that tbh. Your call!

Reply:
The idea is that we can't recover if encoding and decoding fail. Since we control the source data, and how it's been encoded, it is our responsibility to ensure that these must always succeed. If it doesn't, our system goes into an irrecoverable state and it should simply panic.

Please see: https://doc.rust-lang.org/beta/book/ch09-03-to-panic-or-not-to-panic.html#:~:text=The%20panic!,your%20code%20could%20recover%20from.

Reply:
Right, but I still wonder: are we not allowed to error on malformed payload and then simply continue ? One messages might be lost but we wont stop the entire publishing ? Its more of a design decision maybe ?

Reply:
Yes, exactly. We are publishing sensitive data possibly used in DeFi applications, so my design approach so far has been in favor of 100% correctness first.

Reply:
> One messages might be lost but we wont stop the entire publishing ?

We should stop publishing since the stream chains possibly dependent data. We can only decouple concerns like that by `chain_id` since streams from Ethereum are different from Solana for example.